src/System.Net.Primitives/src/System/Net/Sockets/SocketAddress.cs

```diff
@@ -0,0 +1,283 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace System.Net
+{
+    using System;
+    using System.Runtime.InteropServices;
+    using System.Net.Sockets;
+    using System.Text;
+    using System.Globalization;
+    using System.Diagnostics.Contracts;
+
+    /// <devdoc>
+    ///    <para>
+    ///       This class is used when subclassing EndPoint, and provides indication
+    ///       on how to format the memory buffers that winsock uses for network addresses.
+    ///    </para>
+    /// </devdoc>
+    public class SocketAddress
+    {
+        internal const int IPv6AddressSize = 28;
+        internal const int IPv4AddressSize = 16;
+
+        internal int InternalSize;
+        internal byte[] Buffer;
+
+        private const int WriteableOffset = 2;
+        private const int MaxSize = 32; // IrDA requires 32 bytes
+        private bool _changed = true;
+        private int _hash;
+
+        public AddressFamily Family
+        {
+            get
+            {
+                int family;
+#if BIGENDIAN
+                family = ((int)Buffer[0]<<8) | Buffer[1];
+#else
+                family = Buffer[0] | ((int)Buffer[1] << 8);
+#endif
+                return (AddressFamily)family;
+            }
+        }
+
+        public int Size
+        {
+            get
+            {
+                return InternalSize;
+            }
+        }
+
+        // Access to unmanaged serialized data. This doesn't
+        // allow access to the first 2 bytes of unmanaged memory
+        // that are supposed to contain the address family which
+        // is readonly.
+        public byte this[int offset]
+        {
+            get
+            {
+                if (offset < 0 || offset >= Size)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+                return Buffer[offset];
+            }
+            set
+            {
+                if (offset < 0 || offset >= Size)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+                if (Buffer[offset] != value)
+                {
+                    _changed = true;
+                }
+                Buffer[offset] = value;
+            }
+        }
+
+        public SocketAddress(AddressFamily family) : this(family, MaxSize)
+        {
+        }
+
+        public SocketAddress(AddressFamily family, int size)
+        {
+            if (size < WriteableOffset)
+            {
+                // It doesn't make sense to create a socket address with less tha
+                // 2 bytes: that's where we store the address family.
+                throw new ArgumentOutOfRangeException("size");
+            }
+            InternalSize = size;
+            Buffer = new byte[(size / IntPtr.Size + 2) * IntPtr.Size];
+
+#if BIGENDIAN
```

Comment:
@stephentoub I think the original dev's intent was to avoid a conditional on Windows where BIGENDIAN should always be false. 

Maybe we can postpone this until we start the x-plat PAL? (The entire BIGENDIAN block should disappear for Windows.)